### DISCUSSION TITLE
[pom-less] Incorrect dependencies in generated pom.xml

### DISCUSSION THREAD

**Discussion by fipro78** [eclipse-tycho/tycho#1573] [Q&A]
Hi,

I noticed that the pom.xml files that are generated via pom-less Tycho now contain a `dependencies` section. I can't tell when this feature was added, but since we updated to Tycho 2.7.0 this is the case. That actually broke the Maven usage of our bundles, as the generated dependencies are incorrect.

```
    <dependency>
      <groupId>p2.eclipse.plugin</groupId>
      <artifactId>javax.annotation</artifactId>
      <version>1.2.0.v201602091430</version>
      <type>eclipse-plugin</type>
      <scope>provided</scope>
      <optional>true</optional>
    </dependency>
```

They are provided and optional and should not break actually, but having incorrect metadata in a published artifact also doesn't feel right.

More critical is the change of the parent. In current Tycho versions the parent in a pom-less generated pom.xml looks like this:

```
  <parent>
    <groupId>org.eclipse.app4mc.migration</groupId>
    <artifactId>plugins</artifactId>
    <version>2.1.0</version>
  </parent>
```

Previously it had a relativePath:

```
  <parent>
    <groupId>org.eclipse.app4mc.migration</groupId>
    <artifactId>plugins</artifactId>
    <version>2.0.0</version>
    <relativePath>../.polyglot.pom.tycho</relativePath>
  </parent>
```

With the removal of the relativePath Maven expects that pom needs to be published. But as it does not really exist (is that pom.xml really generated as persistent artifact?) the builds are now all failing because of the missing parent.

Is there any workaround for this? Or how can we solve the Maven Central publishing with the pom-less Tycho generated artefacts?

**Comment by laeubi** [comment]
First of all, please use Tycho 3.0 as 2.x is kind of "legacy" now, and even for 2.x the latest release is 2.7.5.

> I noticed that the pom.xml files that are generated via pom-less Tycho now contain a `dependencies` section. 

This is not done by tycho-pomless but the tycho consumer pom feature:

https://tycho.eclipseprojects.io/doc/3.0.0/tycho-packaging-plugin/update-consumer-pom-mojo.html

> With the removal of the relativePath Maven expects that pom needs to be published. But as it does not really exist (is that pom.xml really generated as persistent artifact?) the builds are now all failing because of the missing parent.

Yes "pomless" don't mean there are no poms but that they are generated on the fly.

> Is there any workaround for this? Or how can we solve the Maven Central publishing with the pom-less Tycho generated artefacts?

Use a more recent tycho, or disable the pom generation in your parent pom:

https://tycho.eclipseprojects.io/doc/3.0.0/tycho-packaging-plugin/update-consumer-pom-mojo.html#skipPomGeneration

**Reply by fipro78** [reply]
Ok, seems to be a way to go. Although, Tycho 3.0 requires Java 17 for the execution right?

For the parent issue I noticed that the mistake was done on our side. I forgot to specify tycho.pomless.parent in the build.properties when I removed the pom.xml files per plugin project. :(

Should skipPomGeneration be set as build property in the pom.xml or is it necessary to configure a plugin somehow?

**Reply by laeubi** [reply]
> Ok, seems to be a way to go. Although, Tycho 3.0 requires Java 17 for the execution right?

Yes you need java 17 to run the build, but camn still use older java to compile.

> For the parent issue I noticed that the mistake was done on our side. I forgot to specify tycho.pomless.parent in the build.properties when I removed the pom.xml files per plugin project. :(

> Should skipPomGeneration be set as build property in the pom.xml or is it necessary to configure a plugin somehow?

You configure it in the plugin settings e.g.:

https://github.com/eclipse-tycho/tycho/blob/fb877cb750f917636de0d2b942248955a384a908/tycho-bundles/pom.xml#L43-L50
